DataOnion/Auth/SlidingExpirationStrategy.cs

```diff
@@ -0,0 +1,105 @@
+using StackExchange.Redis;
+
+namespace DataOnion.Auth
+{
+    public class SlidingExpirationStrategy<T> : IAuthServiceStrategy<T>
+        where T : class, IAuthStorable<T>
+    {
+        public static string ExpiryKey { get; } = "expiration";
+
+        private readonly IDatabase _database;
+        private readonly TimeSpan _timeout;
+        private readonly Func<HashEntry[], T> _makeFromHash;
+
+        private string _expirationTimeStr => DateTimeOffset.UtcNow
+            .Add(_timeout)
+            .ToUnixTimeSeconds()
+            .ToString();
+
+        public SlidingExpirationStrategy(
+            IDatabase database,
+            TimeSpan timeout,
+            Func<HashEntry[], T> makeFromHash
+        )
+        {
+            _database = database;
+            _timeout = timeout;
+            _makeFromHash = makeFromHash;
+        }
+
+        private async Task SetExpirationAsync(RedisKey id)
+        {
+            await _database.HashSetAsync(
+                id,
+                ExpiryKey,
+                _expirationTimeStr
+            );
+        }
+
+        public async Task<bool> LoginAsync(T details)
+        {
+            var id = new RedisKey(details.GetId());
+            var redisHash = await _database.HashGetAllAsync(id);
+
+            if (redisHash.Length == 0)
+            {
+                await _database.HashSetAsync(
+                    id,
+                    details
+                        .ToRedisHash()
+                        .Append(new HashEntry(ExpiryKey, _expirationTimeStr))
+                        .ToArray()
+                );
+                return true;
+            }
+            else
+            {
+                var session = _makeFromHash(redisHash);
+                if (details.IsSameSessionAs(session))
+                {
+                    await SetExpirationAsync(id);
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public async Task<T?> CheckSessionAsync(string id)
+        {
+            var redisKey = new RedisKey(id);
+            var expirationRedisValue = await _database.HashGetAsync(
+                redisKey,
+                ExpiryKey
+            );
```

Comment:
why wouldnt we just fetch the full hash at the start rather than conditionally fetching it twice?

Reply:
Getting the entire hash is O(n), but getting one value is O(1). Notice that there are some short-circuit paths in which the rest of the hash is not fetched. It's difficult if not impossible to reason about the general case, but at least in the case that the session is invalid, this should be more performant.

Reply:
kk